plugins/hls-selection-range-plugin/src/Ide/Plugin/SelectionRange.hs

```diff
@@ -0,0 +1,107 @@
+import           Development.IDE.GHC.Compat              (HieAST (Node), Span,
+                                                          getAsts)
+import           Development.IDE.GHC.Compat.Util
+import           Ide.Plugin.SelectionRange.ASTPreProcess (PreProcessEnv (PreProcessEnv),
+                                                          preProcessAST)
+import           Ide.Types                               (PluginDescriptor (pluginHandlers),
+                                                          PluginId,
+                                                          defaultPluginDescriptor,
+                                                          mkPluginHandler)
+import           Language.LSP.Server                     (LspM)
+import           Language.LSP.Types                      (List (List),
+                                                          NormalizedFilePath,
+                                                          Position,
+                                                          ResponseError,
+                                                          SMethod (STextDocumentSelectionRange),
+                                                          SelectionRange (..),
+                                                          SelectionRangeParams (..),
+                                                          TextDocumentIdentifier (TextDocumentIdentifier))
+import           Prelude                                 hiding (span)
+
+descriptor :: PluginId -> PluginDescriptor IdeState
+descriptor plId = (defaultPluginDescriptor plId)
+    { pluginHandlers = mkPluginHandler STextDocumentSelectionRange selectionRangeHandler
+    }
+
+selectionRangeHandler :: IdeState -> PluginId -> SelectionRangeParams -> LspM c (Either ResponseError (List SelectionRange))
+selectionRangeHandler ide _ SelectionRangeParams{..} = do
+    let (TextDocumentIdentifier uri) = _textDocument
+    let filePathMaybe = toNormalizedFilePath' <$> uriToFilePath' uri
+    case filePathMaybe of
+        Nothing -> pure . Right . List $ []
+        Just filePath -> liftIO $ do
+            let (List positions) = _positions
+            selectionRanges <- runIdeAction "SelectionRange" (shakeExtras ide) $ getSelectionRanges filePath positions
+            pure . Right . List $ selectionRanges
+
+getSelectionRanges :: NormalizedFilePath -> [Position] -> IdeAction [SelectionRange]
+getSelectionRanges file positions = fmap (fromMaybe []) <$> runMaybeT $ do
+    (HAR{hieAst, refMap}, positionMapping) <- useE GetHieAst file
+    positions' <- MaybeT . pure $ traverse (fromCurrentPosition positionMapping) positions
+    ast <- MaybeT . pure $ getAsts hieAst Map.!? (coerce. mkFastString . fromNormalizedFilePath) file
+    let ast' = runReader (preProcessAST ast) (PreProcessEnv refMap)
+    MaybeT . pure . traverse (toCurrentSelectionRange positionMapping) $
+        findSelectionRangesByPositions (astPathsLeafToRoot ast') positions'
```

Comment:
These two lines in particular are tricky! A comment explaining what the idea is here would help